xarray/core/missing.py

```diff
@@ -332,3 +369,168 @@ def _get_valid_fill_mask(arr, dim, limit):
+
+
+def _assert_single_chunk(var, axes):
+    for axis in axes:
+        if len(var.chunks[axis]) > 1 or var.chunks[axis][0] < var.shape[axis]:
+            raise NotImplementedError(
+                'Chunking along the dimension to be interpolated '
+                '({}) is not supported.'.format(axis))
+
+
+def _localize(var, indexes_coords):
+    """ Speed up for linear and nearest neighbor method.
+    Only consider a subspace that is needed for the interpolation
+    """
+    indexes = {}
+    for dim, [x, new_x] in indexes_coords.items():
+        index = x.to_index()
+        imin = index.get_loc(np.min(new_x.values), method='nearest')
+        imax = index.get_loc(np.max(new_x.values), method='nearest')
+
+        indexes[dim] = slice(max(imin - 2, 0), imax + 2)
+        indexes_coords[dim] = (x[indexes[dim]], new_x)
+    return var.isel(**indexes), indexes_coords
+
+
+def interp(var, indexes_coords, method, **kwargs):
+    """ Make an interpolation of Variable
+
+    Parameters
+    ----------
+    var: Variable
+    index_coords:
+        Mapping from dimension name to a pair of original and new coordinates.
+        Original coordinates should be sorted in strictly ascending order.
+        Note that all the coordinates should be Variable objects.
+    method: string
+        One of {'linear', 'nearest', 'zero', 'slinear', 'quadratic',
+        'cubic'}. For multidimensional interpolation, only
+        {'linear', 'nearest'} can be used.
+    **kwargs:
+        keyword arguments to be passed to scipy.interpolate
+
+    Returns
+    -------
+    Interpolated Variable
+
+    See Also
+    --------
+    DataArray.interp
+    Dataset.interp
+    """
+    if not indexes_coords:
+        return var.copy()
+
+    # simple speed up for the local interpolation
+    if method in ['linear', 'nearest']:
+        var, indexes_coords = _localize(var, indexes_coords)
+
+    # default behavior
+    kwargs['bounds_error'] = kwargs.get('bounds_error', False)
+
+    # target dimensions
+    dims = list(indexes_coords)
+    x, new_x = zip(*[indexes_coords[d] for d in dims])
+    destination = broadcast_variables(*new_x)
+
+    # transpose to make the interpolated axis to the last position
+    broadcast_dims = [d for d in var.dims if d not in dims]
+    original_dims = broadcast_dims + dims
+    new_dims = broadcast_dims + list(destination[0].dims)
+    interped = interp_func(var.transpose(*original_dims).data,
+                           x, destination, method, kwargs)
+
+    result = Variable(new_dims, interped, attrs=var.attrs)
+
+    # transpose the result so that the new dimensions are inserted to the
+    # original position
+    if all(x1.dims == new_x1.dims for x1, new_x1 in zip(x, new_x)):
+        result_dims = var.dims
+    else:
+        start = min(var.get_axis_num(d) for d in dims)
+        result_dims = (broadcast_dims[:start] + list(destination[0].dims) +
+                       broadcast_dims[start:])
```

Comment:
If I recall correctly, `Variable.__getitem__` orders dimensions based on broadcasting:
- Unindexed dimensions are assumed to be indexed with `slice(None)`
- `slice()` is equivalent to `range(...)`
- Unlabeled arrays are equivalent to indexing along the matching dimension
- Dimensions on the broadcast result appear in order of appearance.

I'm not confident this version is correct in all edge cases, so I think it would be best to match this logic directly, e.g., by adding dimensions from all variables to an `OrderedSet` (from `xarray.utils`).

Reply:
OK. Done. Thanks for the suggestion.